sapi/phpdbg/phpdbg_rinit_hook.c

```diff
@@ -67,11 +67,14 @@ static PHP_RINIT_FUNCTION(phpdbg_webhelper) /* {{{ */
 		}
 
 		char *msg = NULL;
-		char msglen[5] = {0};
-		phpdbg_webdata_compress(&msg, (int *)msglen);
-
-		send(s, msglen, 4, 0);
-		send(s, msg, *(int *) msglen, 0);
+		size_t msglen = 0;
+		phpdbg_webdata_compress(&msg, &msglen);
+
+		send(s, (unsigned char[]){ msglen
+			                     , msglen / 0x100
+			                     , msglen / 0x10000
+			                     , msglen / 0x1000000 }, 4, 0);
```

Comment:
This looks like a genuine bug in the original code, in that it causes a dependence on the machine endianness, which I presume was not intended.

@bwoebi @krakjoe I'm not familiar with where this protocol is used. Is it correct that the length should be encoded as little-endian?

Reply:
Btw, this should be moved into an explicit declaration, as compound literals are not C89. Not that this file appears to conform to C89 anyway, I guess it never came up due to the WIN32 guard.

Reply:
@nikic The most common representation of integers is little endian, and that's the native x86 architecture. I'm about 99.9% certain that's what was intended. It usually is someone thinking they're doing something clever by manually optimising what our compilers could've easily done even fifteen or twenty years ago.

It's probably a good idea to extract that _serialisation_ logic into a function to be reused anywhere else such a serialisation might be used (and a deserialisation function, which is the other part of the bug I haven't checked out, should also be written). After all, that's the reason we choose *named* objects (i.e. variables) rather than *unnamed ones* (i.e. my compound literal), right? Because we're going to reuse that identifier later? Otherwise, the identifier we're only going to use once is just noisy... I mean, if you're going to multiply by some literal value, you don't necessarily want to have to assign that literal value to a variable first, right? I think the patch I just committed should be a reasonable compromise, for now.

I'd suggest an interface similar to `snprintf` and `sscanf`, for the sake of familiarity. The return value being the number of bytes written in the case of the *print* function, and the number of items assigned to in the case of the *scan* function. The only necessary difference is that `%d` etc parse/create a *little endian base 256* representation, as opposed to the *big endian base 10*  representation which the analogous standard functions typically output. It's a shame you can't use C11, because there are some generics available which might allow developers to avoid that pesky *format string* which seems to become out of sync with its parameters so easily... It's an even bigger shame you can't use C++, because generics are even nicer in C++. We could work towards that by writing one small step at a time, as Rasmus did, except following the specs to ensure we make it as *familiar* as possible.

I'm honestly getting sick of one-line functions which merely wrap a macro. There's only one justification for all of this... premature optimisation... Eh, once this is done, I'm outie. I'll pop some food for thought in this review later today, so keep an eye on this space :)